src/tool/hpcrun/gpu/intel/opencl-intercept.c

```diff
@@ -0,0 +1,125 @@
+#include <gotcha/gotcha.h>
+#include <stdio.h>
+#include <CL/cl.h>
+
+#include <hpcrun/messages/messages.h>
+#include "opencl-api.h"
+#include "opencl-intercept.h"
+
+// function declarations
+typedef cl_command_queue (*clqueue_fptr)(cl_context, cl_device_id, cl_command_queue_properties, cl_int *);
+typedef cl_int (*clkernel_fptr)(cl_command_queue, cl_kernel, cl_uint, const size_t*, const size_t*, const size_t*, cl_uint, const cl_event*, cl_event*);
+typedef cl_int (*clreadbuffer_fptr)(cl_command_queue, cl_mem, cl_bool, size_t, size_t, void *, cl_uint, const cl_event *, cl_event *);
+typedef cl_int (*clwritebuffer_fptr)(cl_command_queue, cl_mem, cl_bool, size_t, size_t, const void *, cl_uint, const cl_event *, cl_event *);
+typedef cl_int (*clSetKernelArg_fptr)(cl_kernel, cl_uint, size_t, const void *);
+
+cl_event* eventNullCheck(cl_event*);
+static cl_command_queue clCreateCommandQueue_wrapper(cl_context, cl_device_id, cl_command_queue_properties, cl_int *);
+static cl_int clEnqueueReadBuffer_wrapper(cl_command_queue, cl_mem, cl_bool, size_t, size_t, void *, cl_uint, const cl_event *, cl_event *);
+static cl_int clEnqueueWriteBuffer_wrapper(cl_command_queue, cl_mem, cl_bool, size_t, size_t, const void *, cl_uint, const cl_event *, cl_event *);
+static cl_int clEnqueueNDRangeKernel_wrapper(cl_command_queue, cl_kernel, cl_uint, const size_t*, const size_t*, const size_t*,																		   cl_uint, const cl_event*, cl_event*);
+
+// global variables
+static gotcha_wrappee_handle_t clCreateCommandQueue_handle;
+static gotcha_wrappee_handle_t clEnqueueNDRangeKernel_handle;
+static gotcha_wrappee_handle_t clEnqueueReadBuffer_handle;
+static gotcha_wrappee_handle_t clEnqueueWriteBuffer_handle;
+uint64_t correlation_id = 0;
+
+static cl_command_queue clCreateCommandQueue_wrapper(cl_context context, cl_device_id device, cl_command_queue_properties properties, cl_int *errcode_ret)
+{
+  properties |= (cl_command_queue_properties)CL_QUEUE_PROFILING_ENABLE; // enabling profiling
+  clqueue_fptr clCreateCommandQueue_wrappee = (clqueue_fptr) gotcha_get_wrappee(clCreateCommandQueue_handle);
+  return clCreateCommandQueue_wrappee(context, device, properties, errcode_ret);
+}
+
+static cl_int clEnqueueNDRangeKernel_wrapper(cl_command_queue command_queue, cl_kernel ocl_kernel, cl_uint work_dim, const size_t *global_work_offset, 													   const size_t *global_work_size, const size_t *local_work_size, cl_uint num_events_in_wait_list, 															 const cl_event *event_wait_list, cl_event *event)
+{
+  uint64_t local_correlation_id = __atomic_fetch_add(&correlation_id, 1, __ATOMIC_SEQ_CST);
+  cl_kernel_callback *kernel_cb = (cl_kernel_callback*) malloc(sizeof(cl_kernel_callback));
```

Comment:
I am worried about this stuff. I mentioned this before, we need memory reclamation for this datum (`kernel_cb`).

Reply:
Down the callstack, we are calling free() on this heap object. https://github.com/HPCToolkit/hpctoolkit/blob/3072b5c81773013c43a3962516f3b9b202b9aaf6/src/tool/hpcrun/gpu/intel/opencl-api.c#L74